casper/src/main/scala/coop/rchain/casper/Validate.scala

```diff
@@ -141,6 +151,49 @@ object Validate {
       }
     }
 
+  /**
+    * Validate no deploy by the same (user, millisecond timestamp)
+    * has been produced in the chain
+    */
+  def repeatDeploy[F[_]: Monad: Log: BlockStore](block: BlockMessage,
+                                                 genesis: BlockMessage,
+                                                 dag: BlockDag): F[Boolean] =
+    BlockStore[F].asMap().flatMap { internalMap: Map[BlockHash, BlockMessage] =>
+      {
+        def parents(b: BlockMessage): Iterator[BlockMessage] =
+          ProtoUtil.parents(b).iterator.flatMap(internalMap.get)
+
+        val deployKeySet = block.getBody.newCode.map(d => (d.getRaw.user, d.getRaw.timestamp)).toSet
+        val iterator = DagOperations.bfTraverse(Some(genesis))(b =>
```

Comment:
The algorithm you have seems unnecessary complicated and expensive. You accumulate all of instances of the deploys in the current block which have appeared anywhere in the DAG and then check if those repetitions are in a different fork. Save yourself the trouble and only look on the current fork to begin with.
```scala
val invalid =  bfTraverse[BlockMessage](Some(block))(parents).exists(
  _.body.exists(
       _.newCode.exists(
         _.raw.exists(p => deployKeySet.contains((p.user, p.timestamp)))
      )
  )
)
```

Reply:
Fantastic. And we should exclude current block in traversing.